Add AdGuard DNS Client information (#212)
AdguardTeam/KnowledgeBaseDNS: docs/dns-client/configuration.md

---
title: Configuration file
sidebar_position: 2
---

See file [`config.dist.yml`][dist] for a full example of a [YAML][yaml] configuration file with comments.

<!--
    TODO(a.garipov): Find ways to add IDs to individual list items.
 -->

[dist]: https://github.com/AdguardTeam/AdGuardDNSClient/blob/master/config.dist.yaml
[yaml]: https://yaml.org/

## `dns` {#dns}

> The `dns` object configures the behavior of DNS server. It has the following properties:

Reviewer: ```suggestion
The `dns` object configures the behavior of the DNS server. It has the following properties:
```

Reviewer: а тут будут проперти?

Author: Не до конца понял, о чём ты?

Reviewer: после этой фразы ждешь списка, а по факту просто новые подзаголовки, просто следующего уровня. кажется, что что-то пропущено

Reviewer: может вытащить эти названия, оформить списком, а дальше уже расписать подробнее с подзаголовками?

Author: Я бы так и делал, но у вас нельзя давать ссылки на элементы списка.  А ссылки на эти свойства понадобятся.

### `cache` {#dns-cache}

The `cache` object configures caching the results of querying DNS. It has the following properties:

- `enabled`: Whether or not the DNS results should be cached.

    **Example:** `true`

> - `size`: The maximum size of the DNS results cache as human-readable data size. It must be greater than zero if `enabled` is `true`.

Reviewer: ```suggestion
- `size`: The maximum size of the DNS results cache in human-readable data size. It must be greater than zero if `enabled` is `true`.
```

    **Example:** `128MB`

> - `client_size`: The maximum size of the DNS results cache for each configured client’s address or subnetwork as human-readable data size. It must be greater than zero if `enabled` is `true`.

Reviewer: ```suggestion
- `client_size`: The maximum size of the DNS result cache for each configured client address or subnetwork in human-readable data size. It must be greater than zero if `enabled` is `true`.
```

    **Example:** `4MB`

### `server` {#dns-server}

The `server` object configures the handling of incoming requests. It has the following properties:

- `listen_addresses`: The set of addresses with ports to listen on.

    **Property example:**

    ```yaml
    'listen_addresses':
      - address: '127.0.0.1:53'
      - address: '[::1]:53'
    ```

### `bootstrap` {#dns-bootstrap}

> The `bootstrap` object configures the resolving of [upstream](#dns-upstream) servers addresses. It has the following properties:

Reviewer: ```suggestion
The `bootstrap` object configures the resolving of [upstream](#dns-upstream) server addresses. It has the following properties:
```

Reviewer: + resolving → resolution?

> - `servers`: The list of servers to use for resolving upstream servers hostnames.

Reviewer: ```suggestion
- `servers`: The list of servers to resolve the hostnames of upstream servers.
```

    **Property example:**

    ```yaml
    'servers':
      - address: '8.8.8.8:53'
      - address: '192.168.1.1:53'
    ```

- `timeout`: The timeout for bootstrap DNS requests as a human-readable duration.

    **Example:** `2s`

### `upstream` {#dns-upstream}

The `upstream` object configures the actual resolving of requests. It has the following properties:

- `groups`: The set of upstream servers keyed by the group’s name. It has the following fields:

    - `address`: The upstream server’s address.

        **Example:** `'8.8.8.8:53'`

    - `match`: The list of criteria to match the request against. Each entry may contain the following properties:

>         - `question_domain`: The domain or a suffix of the domain that the set of upstream servers should be used to resolve.

Author: This section causes issues with `docusaurus` 2.4.3 on my machine.  I wasn't able to find a way to make both `docusaurus` and `markdownlint` happy about it.

            **Example:** `'mycompany.local'`

>         - `client`: The client’s address or a subnet of the client’s address that the set of upstream servers should be used to resolve requests from. It must have no significant bits outside of the subnet mask.

Reviewer: ```suggestion
        - `client`: The client’s address or a subnet of the client’s address from which the set of upstream servers should resolve requests. It must have no significant bits outside the subnet mask.
```

            **Example:** `'192.0.2.0/24'`

>         :::note

Reviewer 2: проверить не могу, а точно при смещении :::note относительно начала строки ничего не поедет нигде?

        Properties specified within a single entry are combined with a logical AND. Entries are combined with a logical OR.

        :::

        **Property example:**

        ```yaml
        'match':
          - question_domain: 'mycompany.local'
            client: '192.168.1.0/24'
          - question_domain: 'mycompany.external'
          - client: '1.2.3.4'
        ```

    :::info

    `groups` should contain at least a single entry named `default`, and optionally a single entry named `private`, both should have no `match` property.

    :::

    The `default` group will be used when there are no matches among other groups. The `private` group will be used to resolve the PTR requests for the private IP addresses. Such queries will be answered with `NXDOMAIN` if no `private` group is defined.

- `timeout`: The timeout for upstream DNS requests as a human-readable duration.

    **Example:** `2s`

### `fallback` {#dns-fallback}

> The `fallback` object configures the behavior of DNS server on failures. It has the following properties:

Reviewer: ```suggestion
The `fallback` object configures the behavior of the DNS server in case of failure. It has the following properties:
```

    **Example:** `2s`

## `debug` {#debug}

The `debug` object configures the debugging features. It has the following properties:

### `pprof` {#debug-pprof}

The `pprof` object configures the [`pprof`][pkg-pprof] HTTP handlers. It has the following properties:

- `port`: The port to listen on for debug HTTP requests on localhost.

    **Example:** `6060`

- `enabled`: Whether or not the debug profiling is enabled.

    **Example:** `true`

[pkg-pprof]: https://golang.org/pkg/net/http/pprof

## `log` {#log}

The `log` object configures the logging. It has the following properties:

- `output`: The output to which logs are written.

    :::note

    Log entries written to the system log are in `text` format (see below) and use the system timestamp.

    :::

    Possible values:

    - `syslog` means that the platform-specific system log is used, which is syslog for Linux and Event Log for Windows.

    - `stdout` for standard output stream.

    - `stderr` for standard error stream.

    - Absolute path to the log file.

      **Example:** `/home/user/logs`

      **Example:** `C:\Users\user\logs.txt`

>     **Example:** `syslog`

Reviewer: а тут ок с отступом?

Author: Да, так как это пример для всего поля, а те, что выше, для одного из типов.

> - `format`: Specifies format of the log entries.

Reviewer: ```suggestion
- `format`: Specifies the format of the log entries.
```

    Possible values:

>     - `adguard_legacy`;

Reviewer: мы вроде пишем без знаков препинания в таких случаях. будем убирать или не надо?

Author: Не пишем знаков в списках?

Reviewer: ага

    - `default`;
    - `json`;
    - `jsonhybrid`;
    - `text`.

    **Example:** `default`

    <!--
        TODO(s.chzhen):  Add output examples.
    -->

> - `timestamp`: Specifies if the log entries should contain timestamp.

Reviewer: ```suggestion
- `timestamp`: Specifies whether to include a timestamp in the log entries.
```

    **Example:** `false`

> - `verbose`: Specifies if the log should be more informative.

Reviewer: ```suggestion
- `verbose`: Specifies whether the log should be more informative.
```
